lib/collection/src/collection/mod.rs

```diff
@@ -436,64 +437,57 @@ impl Collection {
-        // is needed.
-        let is_resharding = matches!(
-            current_state,
-            Some(ReplicaState::Resharding | ReplicaState::ReshardingScaleDown)
-        );
-
-        // If *any* of re-sharding replicas are dead, we need to abort re-sharding before marking them as Dead
-        if is_resharding && new_state == ReplicaState::Dead {
-            drop(shard_holder); // need to release lock so abort_resharding() can take write lock
-
-            if let Some(state) = self.resharding_state().await {
-                self.abort_resharding(state.key(), false).await?;
-            }
-
-            // Actually mark as Dead
-            {
-                let shard_holder = self.shards_holder.read().await;
-                let replica_set = shard_holder
-                    .get_shard(shard_id)
-                    .ok_or_else(|| shard_not_found_error(shard_id))?;
-
-                replica_set
-                    .ensure_replica_with_state(peer_id, ReplicaState::Dead)
-                    .await?;
-            }
-
-            return Ok(());
-        }
-
         // Update replica status
         replica_set
             .ensure_replica_with_state(peer_id, new_state)
             .await?;
 
         if new_state == ReplicaState::Dead {
-            // TODO(resharding): Abort all resharding transfers!?
-
-            // Terminate transfer if source or target replicas are now dead
+            let resharding_state = shard_holder.resharding_state.read().clone();
             let related_transfers = shard_holder.get_related_transfers(shard_id, peer_id);
 
-            // `abort_shard_transfer` locks `shard_holder`!
+            // Functions below lock `shard_holder`!
             drop(shard_holder);
 
+            let mut abort_resharding_result = CollectionResult::Ok(());
+            let mut abort_transfers_result = CollectionResult::Ok(());
+
+            // Terminate resharding if we kill a related shard
+            let has_resharding_state = current_state
+                .as_ref()
+                .is_some_and(ReplicaState::is_resharding);
+            match (has_resharding_state, resharding_state) {
+                // Not resharding, nothing to abort
+                (_, None) => {}
+                // If resharding up, abort if replica is in resharding state
+                (true, Some(state)) if state.direction == ReshardingDirection::Up => {
+                    abort_resharding_result = self.abort_resharding(state.key(), false).await;
+                }
+                // If resharding up, abort if we kill target shard
+                (_, Some(state))
+                    if state.direction == ReshardingDirection::Up
+                        && state.peer_id == peer_id
+                        && state.shard_id == shard_id =>
+                {
+                    abort_resharding_result = self.abort_resharding(state.key(), false).await;
+                }
```

Comment:
This is kinda wrong. We have to abort if target shard is in `Resharding` state, but we don't need to abort once it's `Active`. (Especially if we already replicated target shard.)

Reply:
IMO, previous logic, when we relied purely on `Resharding`/`ReshardingScaleDown` was fine. Works the same but easier to read and understand.